### ISSUE TITLE
advanced search: add field not working after going in "back in history"

### ISSUE THREAD

**Issue by akohlbecker** [gbhl/bhl-europe#259]
After going back to advanced search from the result list (via browser back button) it is not possible to add a new search field unless the page has been reloaded.

**Comment by wkollernhm** [comment]
This is a drupal7 problem with AJAX - we won't be able to fix that.

See https://drupal.org/node/1237138

**Comment by akohlbecker** [comment]
is this request for help really applicable to our situation? 

The "add" button is a submit element not a select element. Ok, we are having two submit buttons in that form, this may cause problems. If the problem is having two submit buttons, I guess we could also use the form element type "button" for the add button. According to http://api.drupal.org/api/drupal/developer--topics--forms_api_reference.html/7#ajax ajax callbacks should work the same on submit and button elements.

**Comment by wkollernhm** [comment]
Please read the post carefully. The problem has nothing to do with the actual buttons but instead with the fact that after the user clicked submit (which happens when the search is executed) the ajax does not work anymore (most probably due to drupal's built in form protection).

**Comment by akohlbecker** [comment]
Excuse me Wolfgang for insisting on the resolvability of this issue. First of all I have to admit that you are right, the problem actually lies in the form security mechanism. 

I did some investigation: 
1. processing the ajax request is skipped in ajax_get_form() which is invoked by ajax_form_callback(). Yes this is a security feature!
2. the documentation of this function sates the following: _This function is also the canonical example of how to implement #ajax['path']. If processing is required that cannot be accomplished with a callback, re-implement this function and set #ajax['path'] to the enhanced function._
3. the according section in http://api.drupal.org/api/drupal/includes--ajax.inc/group/ajax reads similar: _By default, #ajax uses 'system/ajax' as its path for submission and thus calls ajax_form_callback() and a defined #ajax['callback'] function. However, you may optionally specify a different path to request or a different callback function to invoke,..._

So I am convinced that there is an indented  solution to this problem, not only a hack, don't you think so?

**Comment by wkollernhm** [comment]
Well this would actually require us to re-implement the whole ajax processing in ajax_get_form.

However if you see this as a viable solution, feel free to implement it.

**Comment by wkollernhm** [comment]
I've re-opened the issue and assigned it to the january sub-release, in addition I've removed the label critical.

I will investigate on this after the november release. As stated above, if you already know a viable solution feel free to implement it.

**Comment by akohlbecker** [comment]
not a complete solution but I can offer some more or less working custom implementation of ajax_get_form(). It works but it always adds  two fields. I guess we can do something about it where I added the comment _//TODO avoid duplicate execution here?_

``` php
function bhle_search_ajax_get_form() {
    $form_state = form_state_defaults();

    $form_build_id = $_POST['form_build_id'];

    // Get the form from the cache.
    $form = form_get_cache($form_build_id, $form_state);

    if (!$form) {
        $form_id = $_POST['form_id'];
        $form = drupal_build_form($form_id, $form_state);
                                            //TODO avoid duplicate execution here?
    }

    // Since some of the submit handlers are run, redirects need to be disabled.
    $form_state['no_redirect'] = TRUE;

    // When a form is rebuilt after Ajax processing, its #build_id and #action
    // should not change.
    // @see drupal_rebuild_form()
    $form_state['rebuild_info']['copy']['#build_id'] = TRUE;
    $form_state['rebuild_info']['copy']['#action'] = TRUE;

    // The form needs to be processed; prepare for that by setting a few internal
    // variables.
    $form_state['input'] = $_POST;
    $form_id = $form['#form_id'];

    return array($form, $form_state, $form_id, $form_build_id);
}
```

**Comment by wkollernhm** [comment]
I will investigate on this tomorrow. I guess it has something to do with the fact that the form is first build and then rebuild....

**Comment by JiriFrank** [comment]
fixed and testable.

**Comment by JiriFrank** [comment]
Ready to be closed.
